packages/pyright-internal/src/analyzer/checker.ts

```diff
@@ -7654,6 +7657,24 @@ export class Checker extends ParseTreeWalker {
         }
     }
 
+    // Warn when `self` or `cls` parameter has a default value
+    private _disallowDefaultInClsSelfParam(node: FunctionNode) {
+        if (node.d.params.length < 1) {
+            return;
+        }
+
+        const { defaultValue, name } = node.d.params[0].d;
+        if (name && defaultValue) {
+            this._evaluator.addDiagnostic(
+                DiagnosticRule.reportGeneralTypeIssues,
```

Comment:
When adding new errors I like to make it a new diagnostic rule where possible so people can turn it off if they want